<div class="alert alert-info">
Привет, Елена! Меня зовут Светлана Чих и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

<div class="alert alert-success">
<b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
</div>
<div class="alert alert-warning">
<b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
</div>
<div class="alert alert-danger">
<b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
</div>
<div class="alert alert-info">
<b>👂 Совет:</b> Какие-то дополнительные материалы
</div>
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
 </div>

# SQL
## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![jupyter](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2)

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], 
                                                         db_config['host'], db_config['port'], db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Откроем по очереди все таблицы

In [25]:
query = '''

SELECT *
FROM books

LIMIT 5
'''

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
query = '''

SELECT *
FROM authors
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
query = '''

SELECT *
FROM publishers
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
query = '''

SELECT *
FROM ratings
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [16]:
query = '''

SELECT *
FROM reviews
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-success">
<b>👍 Успех:</b> Есть описание задачи, схема БД, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
</div>

## Выполним задания

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [23]:
query = '''

SELECT COUNT(DISTINCT book_id)
FROM books
WHERE publication_date >= '2000-01-01'

'''

pd.io.sql.read_sql(query, con = engine)

count
0    821

Всего с 1 января 2000 года вышло 821 книга. 

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

###  Для каждой книги посчитайте количество обзоров и среднюю оценку

In [21]:
query = '''

SELECT b.book_id, 
       b.title,
       COUNT(DISTINCT re.review_id) AS cnt_reviews,
       ROUND(AVG(rating),2) AS avg_rating
FROM books AS b
LEFT JOIN reviews AS re ON b.book_id = re.book_id
JOIN ratings AS rat ON b.book_id = rat.book_id

GROUP BY b.book_id, b.title
ORDER BY cnt_reviews DESC
'''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  cnt_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        963                                Water for Elephants            6   
2        734                                   The Glass Castle            6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
4        695  The Curious Incident of the Dog in the Night-Time            6   
..       ...                                                ...          ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
996      808                            The Natural Way to Draw            0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
998      221                          Essential Tales and Poems            0   
999      191  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 4 columns]

База насчитывает 1000 книг, 994 книги имеют хотя бы 1 обзор.

<div class="alert alert-danger">
<s><b>😔 Необходимо исправить:</b> В итоговую таблицу должны войти все книги, даже если у них нет обзоров</s>
</div>

<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Поправила.</div>

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

###  Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [22]:
query = '''

SELECT p.publisher,
       b.publisher_id 
       
FROM books AS b 
JOIN publishers AS p ON p.publisher_id = b.publisher_id

WHERE b.num_pages > 50

GROUP BY b.publisher_id, p.publisher
ORDER BY COUNT(b.book_id) DESC

LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

publisher  publisher_id
0  Penguin Books           212

Издание Penguin Books выпустило наобильшее колличество книг, 212 новых книг.

<div class="alert alert-danger">
<s><b>😔 Необходимо исправить:</b> временная таблица здесь избыточна, попробуй посчитать количество выпущенных книг из объединения таблиц books и publishers (ответ будет другой и книг выпущенно меньше)</s>
</div>

<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Поправила.</div>

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

###  Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [23]:
query = '''

WITH top AS
    (SELECT book_id, 
            COUNT(rating) As count_raiting
    FROM ratings

    GROUP BY book_id
    HAVING COUNT(rating) >=50)
    
SELECT a.author, AVG(r.rating) AS raiting
FROM books b
JOIN ratings AS r ON b.book_id = r.book_id
JOIN authors AS a ON b.author_id = a.author_id
JOIN top AS t ON b.book_id = t.book_id

GROUP BY a.author
ORDER BY raiting DESC

LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

author   raiting
0  J.K. Rowling/Mary GrandPré  4.287097

Самая высокая средняя оценка достается книгам Дж. Роулинг =)
Учитывались только популярные книги, имеющие более 50 оценок пользователей.

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

###  Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [24]:
query = '''
    
SELECT ROUND(AVG(review_cnt),2)
FROM (SELECT username, 
              COUNT(review_id) AS review_cnt
       FROM reviews 
       WHERE username IN (SELECT username
                          FROM ratings
                          GROUP BY username
                          HAVING COUNT(rating_id) > 50)
       GROUP BY  username) AS rereview_count

'''
pd.io.sql.read_sql(query, con = engine)

round
0  24.33

Активные пользователи в среднем написали 24 обзора.

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

## Вывод

- База насчитывает 1000 книг, 994 книги имеют хотя бы 1 обзор. 
- Всего с 1 января 2000 года вышла новая 821 книга.
- Издание Penguin Books выпустило наобильшее колличество книг, 212 новых книг.
- Самая высокая средняя оценка достается книгам Дж. Роулинг =) Учитывались только популярные книги, имеющие более 50 оценок пользователей.
- Активные пользователи в среднем написали 24 обзора.

<div class="alert alert-success">
<b>👍 Успех:</b> Хорошая работа, молодец! Поздравляю с завершением курса! У тебя хорошо получается работать с даными, и чем больше будет опыта, тем проще тебе будет это делать, но и сейчас у тебя есть много ценных для аналитика качеств, ты умеешь рассуждать и делать выводы, понимаешь как извлечь из сухих данных целые истории! Удачи тебе в будущей карьере!
</div>